mono/metadata/sre.c

```diff
@@ -1547,7 +1549,21 @@ mono_reflection_type_get_handle (MonoReflectionType* ref, MonoError *error)
 				return NULL;
 			}
 		}
-
+		/* Need to resolve the generic_type in order for it to create its generic context. */
```

Comment:
Yep, unfortunately it works on Windows. Actually on windows even `var f = r.GetField("sr")` succeeds and you only get a TLE when you do `f.GetType()` (or `f.ToString()` or anything else that actually needs to work with the type `S`).  All without a CreateType for the S builder.

On Mono `var f = r.GetField("sr")` will throw a TLE with my changes, so we're still not as lazy as .NET.  But this is enough to handle the Rhino-Mock case from the [Bugzilla report](https://bugzilla.xamarin.com/show_bug.cgi?id=47867).

But I think the common situation is that you want to build a forest of interdependent types and you don't know the order that `CreateType()` should be called in.  (And with generics you could get into a situation where there's not a strict order).

The real problem is that `reflection_setup_internal_class` sets `klass->class_kind = MONO_CLASS_DEF` unconditionally and later we fix it up to `MONO_CLASS_GTD` when we can't avoid setting up a generic context.   Potentially we could go the other way and set it up initially as a `MONO_CLASS_GTD` and then downgrade it to a `MONO_CLASS_DEF` when we're finishing up construction and we see that no type variables were added.

But really this is just the loader problem exasperated - we need a way to deal with partially loaded/constructed types because they could be referenced from other types.  In SRE they just happen to live much longer than in the loader and user code is running.